src/IR/LambdaLift.hs

```diff
@@ -0,0 +1,205 @@
+{-# LANGUAGE DerivingVia #-}
+module IR.LambdaLift where
+
+import qualified Common.Compiler               as Compiler
+import           Common.Identifiers
+import qualified IR.IR                         as I
+
+import qualified IR.Types.Poly                 as Poly
+
+import           Control.Monad.Except           ( MonadError(..) )
+import           Control.Monad.State.Lazy       ( MonadState
+                                                , StateT(..)
+                                                , evalStateT
+                                                , gets
+                                                , modify
+                                                , unless
+                                                )
+
+import qualified Data.Bifunctor                as B
+import qualified Data.Foldable                 as F
+import qualified Data.Map                      as M
+import           Data.Maybe                     ( catMaybes )
+import qualified Data.Set                      as S
+
+data LiftCtx = LiftCtx
+  { globalScope  :: S.Set I.VarId
+  , currentScope :: S.Set I.VarId
+  , freeTypes    :: M.Map I.VarId Poly.Type
```

Comment:
Because I don't need their types. Do you think making them all maps would be more extensible?

Reply:
What is `freeTypes`?

Reply:
Ah sorry I didn't respond to that. Yes, it maps ID to type.

Reply:
Doesn't that mean you need their types, if you're maintaining this separate map?

Reply:
No, I maintain this map because I need the types of the free variables to construct properly typed lambdas

I simply need a set of the names in the current scope so I know when something is/isn't a free variable. I don't care about the types of names in the current scope. Likewise, I carry around the global scope so I can properly descend into new scopes, where the only thing that should be in the new scope is the global scope and possibly some new names.

Reply:
Could you document what each of these members are in the code?